### ISSUE TITLE
[Go SDK] Transform API should accept an Output for the returned property map

### ISSUE THREAD

**Issue by EronWright** [pulumi/pulumi#16756]
Consider changing the transform API to accept an Output for the `Props` (as a whole).  The current approach accepts `pulumi.Map` (a `map[string]Input`), which is difficult to work with, e.g. when transforming inputs asynchronously.  One turns to using `internals.UnsafeAwaitOutput` as a workaround (see the inner `transform` function in the examples), but it would be preferable to simply do:
```go
return &pulumi.ResourceTransformResult{
	Props: rta.Props.ToMapOutputWithContext(ctx).ApplyT(applier),
	Opts:  rta.Opts,
}
```

Here's a couple of examples, showing complex transformations of a Kubernetes object (`metadata`, `spec` as input). Observe how the metadata property is used to adjust the spec property (deeply).
 
```go
applyPatchForceAnnotation := func(ctx context.Context, rta *pulumi.ResourceTransformArgs) *pulumi.ResourceTransformResult {
	transform := func(applier interface{}) {
		o := rta.Props.ToMapOutputWithContext(ctx).ApplyT(applier)
		r, err := internals.UnsafeAwaitOutput(ctx, o)
		if err != nil {
			panic(err)
		}
		rta.Props = r.Value.(pulumi.Map)
	}

	switch rta.Type {
	case "kubernetes:helm.sh/v4:Chart":
		// Do nothing for Helm charts
	default:
		transform(func(obj map[string]any) pulumi.Map {
			// note: obj is an ordinary Unstructured object at this point.
			unstructured.SetNestedField(obj, "true", "metadata", "annotations", "pulumi.com/patchForce")
			return pulumi.ToMap(obj)
		})
	}
	return &pulumi.ResourceTransformResult{
		Props: rta.Props,
		Opts:  rta.Opts,
	}
}

transformImage := func(ctx context.Context, rta *pulumi.ResourceTransformArgs) *pulumi.ResourceTransformResult {
	transform := func(applier interface{}) {
		o := rta.Props.ToMapOutputWithContext(ctx).ApplyT(applier)
		r, err := internals.UnsafeAwaitOutput(ctx, o)
		if err != nil {
			panic(err)
		}
		rta.Props = r.Value.(pulumi.Map)
	}

	switch rta.Type {
	case "kubernetes:apps/v1:Deployment":
		transform(func(obj map[string]any) pulumi.Map {
			// note: obj is an ordinary Unstructured object at this point.

			// adjust the container image based on the labels of the deployment (for demonstration purposes)
			labels, _, _ := unstructured.NestedStringMap(obj, "metadata", "labels")
			if app, ok := labels["app.kubernetes.io/instance"]; ok && app == "ingresscontroller" {
				// change the image field for demonstration purposes
				containers, _, _ := unstructured.NestedFieldNoCopy(obj, "spec", "template", "spec", "containers")
				container := containers.([]any)[0].(map[string]any)
				unstructured.SetNestedField(container, "nginx/nginx-ingress:2.3.1-patched", "image")
			}

			return pulumi.ToMap(obj)
		})
	}

	return &pulumi.ResourceTransformResult{
		Props: rta.Props,
		Opts:  rta.Opts,
	}
}
```

**Comment by Frassle** [comment]
While I agree this is a good example scenario showing that the current `Map` output makes for a difficult function to write, I don't think it's safe to change the return value to an `MapOutput`. 

The issue with an `MapOutput` is the whole value may resolve to `unknown` and there's no way on the engine/provider protocol to indicate that the entire state of a resource is unknown. We always expect to know what the top-level shape of a resource is, i.e. what top level keys are set. In the case where the props resolve to unknown the only value the SDK could send back to the engine is an empty map, and the behaviour of that is definitely not the same as being able to say it's an unknown map.

In this case, I think the best thing to do is to rewrite the apply to use `all` on the 'metadata' and 'spec' keys but to return just the 'spec' map to set the 'spec' property with.

There's probably output combinators we could write to make this easier, especially with all the nesting, and the Go SDK is definitely the most tricky language to work with outputs in due to it's non-generic nature.

**Comment by sklarsa** [comment]
> In this case, I think the best thing to do is to rewrite the apply to use all on the 'metadata' and 'spec' keys but to return just the 'spec' map to set the 'spec' property with.

@Frassle would you mind posting an example of this? I'm having an issue where:
1. I created an object without passing a prop for a `StringPtrInput`
2. I'm updating the object by passing a nil `StringPtrInput`
3. This is causing diffs, even though the value is effectively unchanged

So I would like to set up a transform to obtain the value of the `StringPtrInput` and delete the key from props if it's nil. Also discussing this on Slack: https://pulumi-community.slack.com/archives/CCWP5TJ5U/p1742603552743909

**Comment by Frassle** [comment]
So @sklarsa I think your case is simple because you know your dealing with just nil, not an output value so I think the following will work:
```
pulumi.Transforms([]pulumi.ResourceTransform{
	func(ctx context.Context, rta *pulumi.ResourceTransformArgs) *pulumi.ResourceTransformResult {
		if newProp, ok := rta.Props["kmsKeyId"]; ok && newProp == nil {
			delete(rta.Props, "kmsKeyId")
		}

		return &pulumi.ResourceTransformResult{
			Props: rta.Props,
			Opts:  rta.Opts,
		}
	},
}),
```

**Comment by sklarsa** [comment]
@Frassle I guess I didn't clarify that the reason I changed it to a `StringPtrInput` is because that _may_ have a value in the future. So I still need to inspect the output to see if its value is `nil` or a string.

**Comment by Frassle** [comment]
Yes but you can just check for if its nil then delete it, if it's anything else (which might be an output, or a plain value) then leave it alone.

**Comment by sklarsa** [comment]
This `nil` check won't work in my case, because the new prop has a value of `var kmsKeyId pulumi.StringPtrOutput`.

So I need to check if that output resolves to a `nil` or a `string`. If it resolves to `nil`, I would like no diff because that resolved `nil` value is conceptually the same as not passing anything at all. But if the output resolves to a string value, there should be a diff.

**Comment by Frassle** [comment]
In fully generality that just leads back to the issue above, you can't make the whole outputs map depend on an output value. But the transform will only be called with an output value if the StringPtrOutput resolved to a secret or dependency tracking value. If it resolved to a plain nil then it will just be nil in the output.
